scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,256 @@
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let mut checkpoint = receiver.borrow().clone();
+
+            // We are given timestamp of the beginning of the next read window,
+            // so we have to create a checkpoint with lower timestamp.
+            if checkpoint.0 > Duration::from_millis(1) {
+                checkpoint.0 -= Duration::from_millis(1);
+                checkpoint_maker.make_checkpoint(checkpoint).await?;
+            }
+
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for creating checkpoints.
+#[async_trait]
+pub trait CDCCheckpointMaker: Send + Sync {
+    /// Creates new checkpoint with given timestamp.
+    async fn make_checkpoint(&self, checkpoint: Checkpoint) -> anyhow::Result<()>;
+    /// Loads last seen generation from the checkpoint table.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    /// Loads last seen timestamp for given stream id from the checkpoint table.
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+
+/// Default implementation for CDCCheckpointMaker trait.
+pub struct CDCDefaultCheckpointMaker {
+    session: Arc<Session>,
+    keyspace: String,
+    checkpoint_tablename: String,
+    prepared_stmt: PreparedStatement,
+}
+
+impl CDCDefaultCheckpointMaker {
+    /// Creates new CDCDefaultCheckpointMaker.
+    /// If given None value as table name,
+    /// it uses default value "cdc_checkpoint".
+    pub async fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        tablename: Option<String>,
+    ) -> anyhow::Result<Self> {
+        let session = Arc::clone(session);
+        let checkpoint_tablename = tablename.unwrap_or_else(|| "CDC_CHECKPOINT".to_string());
+
+        CDCDefaultCheckpointMaker::create_table(&session, &keyspace, &checkpoint_tablename).await?;
+        session.await_schema_agreement().await?;
+
+        let prepared_stmt = session
+            .prepare(format!(
+                "UPDATE {}.{} SET time = (?)
+                WHERE stream_id = (?)
+                AND generation = (?)",
+                &keyspace, &checkpoint_tablename
+            ))
+            .await?;
+        let cp_maker = CDCDefaultCheckpointMaker {
+            session,
+            keyspace,
+            checkpoint_tablename,
+            prepared_stmt,
+        };
+
+        Ok(cp_maker)
+    }
+
+    // Creates new table with specified schema in the given keyspace with provided name.
+    // If a table with such name already exists, doesn't do anything.
+    async fn create_table(
+        session: &Session,
+        keyspace: &str,
+        checkpoint_tablename: &str,
+    ) -> anyhow::Result<()> {
```

Comment:
Why not convert it into a method? It already uses most of the `CDCDefaultCheckpointMaker`'s fields.

Reply:
I meant it to be a non-static method, so that it takes `self` and you don't have to explicitly pass `session` and `checkpoint_table`.